src/WtsApi32.Desktop/WtsApi32+WtsSafeMemoryGuardIterator.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) to owners found in https://github.com/AArnott/pinvoke/blob/master/COPYRIGHT.md. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+
+namespace PInvoke
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Runtime.InteropServices;
+    using System.Text;
+
+    /// <content>
+    /// Exported functions from the WtsApi32.dll Windows library
+    /// that are available to Desktop apps only.
+    /// </content>
+    public static partial class WtsApi32
+    {
+        /// <summary>
+        /// Iterator for iterating through memory returned by WtsApi32
+        /// </summary>
+        /// <typeparam name="TIteratorType">Element type</typeparam>
+        internal struct WtsSafeMemoryGuardIterator<TIteratorType> : IEnumerator<TIteratorType>
+        {
+            private readonly IntPtr start;
+            private readonly int underlyingTypeSize;
+            private int index;
+
+            public WtsSafeMemoryGuardIterator(IntPtr start)
+            {
+                this.start = start;
+                this.index = -1;
+                this.underlyingTypeSize = Marshal.SizeOf(typeof(TIteratorType));
+            }
+
+            public TIteratorType Current
+            {
+                get
+                {
+                    return (TIteratorType)Marshal.PtrToStructure(this.start + (this.index * this.underlyingTypeSize), typeof(TIteratorType));
+                }
+            }
+
+            object IEnumerator.Current
+            {
+                get
+                {
+                    return Marshal.PtrToStructure(this.start + (this.index * this.underlyingTypeSize), typeof(TIteratorType));
+                }
+            }
+
+            public void Dispose()
+            {
+            }
+
+            public bool MoveNext()
+            {
+                ++this.index;
```

Comment:
While I see what this class and `WtsSafeSessionInfoGuard` try to achieve the current guidelines of this library are more :
* Expose a the PInvokes as is, in this case with a `out WTS_SESSION_INFO*`.
* Provide helpers for functions with specific memory requirements (In that case it would be a `WTSEnumerateSessions` version taking only a server handle and returning a `IEnumerable<WTS_SESSION_INFO>`